### DISCUSSION TITLE
Perform a fastforward merge to avoid rebuilding same commit on `master`

### DISCUSSION THREAD

**Discussion by blindpirate** [Mergifyio/mergify#2182] [Workflows]
Hi thanks for the great feature mergify provides!

We're a team where many people work on personal branches and merges to `master` when the work finishes. We want to leverage strict merge to implement our pre-test commit, especially avoid rebuilding same commit on master. We would like to have a workflow:

- People work on PR branches, aka. head branch.
- When everything is ok, mergify merges `master` (aka. base branch) to PR branch because of strict merge. Let's call this merge commit `MergeCommitA`.
- Wait CI to pass.
- Now merge PR branch to `master`. Mergify will generate a new merge commit (let's call it `MergeCommitB`).

We want to avoid `MergeCommitB`, i.e. run `git merge --ff` instead of mergify default `git merge --no-ff` because:

- It makes no sense to create `MergeCommitB`. `MergeCommitA` is up-to-date.
- Even with identical contents, `MergeCommitA`/`MergeCommitB` is recognized as different commits by CI thus requires rebuilding everything, this adds unnecessary pressure to CI.

Is there anything I can do to avoid this problem?

Reference: https://docs.mergify.io/actions/merge.html#git-merge-workflow-and-mergify-equivalent-configuration

**Comment by jd** [comment]
👋  @blindpirate!
Happy to have you on board.

I understand what you want, which is basically re-using the "update" commit as the merge commit. That makes sense to us.

Unfortunately, GitHub API does not allow this, which means Mergify would have to do `git push` itself. Doable, but we don't support this case for now. I've added this to our backlog in case that'd interest other people.

Another question would also be: why are you testing `master` with your CI? Since you have `strict` mode enabled, there's really no point in testing the `master` branch anymore in theory.

**Reply by blindpirate** [reply]
Thanks @jd  for the detailed explanation! 

We use `master`'s status as an indicator of CI health. Plus, we occasionally push to `master` directly when the change is tiny or urgent, so we'd like to enable testing on `master`. Anyway, thanks a lot for the answer!

**Comment by sileht** [comment]
Mergify now provides two ways to do this `fast-forward` of pull request commits to `master` branch:
* the queue/merge option `merge_method: fast-forward` for simple queue.
* the queue rules option `queue_branch_merge_method: fast-forward` for advanced queues with speculative checks and batches.
